IanGRap/Squish
# Respawn as block

Contributors: Austin, James

Goal:
Create a system where players have a chance to respawn after they die. 

What Changed:
- BlockDamager.cs
    - Added logic to the Break() function so that the block spawns a player if it is a RespawnBlock and it squished another player before being destroyed
    - In Update, if the block is a RespawnBlock and FallingPathBlocked() is true, then destroy the block
- Spawner.cs
    - Added a list (which functions as a queue) to store the players waiting to be spawned as a block. 
    - When there is a player waiting to respawn, regular block spawning is paused while the Spawner spawns respawn blocks
- MatchManager.cs
    - Added logic for _undoing_ a player death. I.E. added a function to add a player back into _alivePlayers
- PlayerSquishDetector.cs
    - When a player gets squished, if they were squished by a respawn block, tell the respawn block that it killed a player, so that it knows to spawn in a new one.
- PlayerSpawnBeam.cs
    - Contains logic for the beam of light when a player spawns in
- BlockInput.cs
    - Listens for controller input and moves respawn blocks
- CameraTracker.cs
    - Controls when the respawn block gets added to the camera manager

How to Test:
- Load up the match scene and die. You should see a respawn block spawn in after ~3 seconds
- Make sure you can control the respawn block with controllers
- Make sure that respawn blocks can squish players
- Make sure respawn blocks CANNOT push other blocks
- Make sure respawn blocks are destroyed once they stop falling
- Make sure that if a respawn block kills a player, it spawns in a new one of the correct color and controller input
- Change the Respawn Delay value on the Block Spawner. This changes the minimum delay time between player death and when they respawn as a block
- Change the Camera Track Y Coordinate value on the CameraTracker component of the respawn block prefab. This field describes the height at which the camera starts tracking respawn blocks

PRO TIP:
- This feature is way easier to test if you reduce the block spawning rate. Players don't tend to live long enough with the current spawn rate, at least when nobody is controlling them.
